src/platform/qpg/OTAImageProcessorImpl.cpp

```diff
@@ -0,0 +1,210 @@
+    auto * imageProcessor     = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+
+    // running this in a thread so won't block main event loop
+    ChipLogProgress(SoftwareUpdate, "Q: HandlePrepareDownload");
+
+    qvCHIP_OtaEraseArea();
+    qvCHIP_OtaStartWrite();
+
+    imageProcessor->mDownloader->OnPreparedForDownload(CHIP_NO_ERROR);
+}
+
+void OTAImageProcessorImpl::HandleFinalize(intptr_t context)
+{
+    qvCHIP_OtaStatus_t status = qvCHIP_OtaStatusSuccess;
+    auto * imageProcessor     = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    ChipLogProgress(SoftwareUpdate, "Q: HandleFinalize");
+
+    // FIXME - Versions need to be filled in
+    qvCHIP_OtaSetPendingImage(imageProcessor->mSwVer /*swVer*/, imageProcessor->mHwVer /*hwVer*/, qvCHIP_OtaGetAreaStartAddress(),
+                              static_cast<std::uint32_t>(imageProcessor->mParams.downloadedBytes) /*imgSz*/);
+
+    imageProcessor->ReleaseBlock();
+}
+
+void OTAImageProcessorImpl::HandleAbort(intptr_t context)
+{
+    auto * imageProcessor = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        return;
+    }
+
+    // TODO: What do we do when we abort?
+    ChipLogProgress(SoftwareUpdate, "Q: HandleAbort");
+
+    remove(imageProcessor->mParams.imageFile.data());
+    imageProcessor->ReleaseBlock();
+}
+
+void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
+{
+    qvCHIP_OtaStatus_t status = qvCHIP_OtaStatusSuccess;
+    auto * imageProcessor     = reinterpret_cast<OTAImageProcessorImpl *>(context);
+    if (imageProcessor == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "ImageProcessor context is null");
+        return;
+    }
+    else if (imageProcessor->mDownloader == nullptr)
+    {
+        ChipLogError(SoftwareUpdate, "mDownloader is null");
+        return;
+    }
+
+    ChipLogProgress(SoftwareUpdate, "Q: HandleProcessBlock");
+    // TODO: Process block header if any
+
+    status = qvCHIP_OtaWriteChunk(qvCHIP_OtaGetAreaStartAddress() + imageProcessor->mParams.downloadedBytes,
+                                  static_cast<std::uint16_t>(imageProcessor->mBlock.size()),
+                                  reinterpret_cast<std::uint8_t *>(imageProcessor->mBlock.data()));
+
+    if (status != qvCHIP_OtaStatusSuccess)
+    {
+        imageProcessor->mDownloader->EndDownload(CHIP_ERROR_WRITE_FAILED);
+        return;
+    }
+
+    imageProcessor->mParams.downloadedBytes += imageProcessor->mBlock.size();
+    imageProcessor->mDownloader->FetchNextData();
+}
+
+CHIP_ERROR OTAImageProcessorImpl::SetBlock(ByteSpan & block)
```

Comment:
Please note that there is a bug here (comes from the original Linux implementation of this this function). See https://github.com/project-chip/connectedhomeip/issues/13393 . 
There's a plan to fix that issue in all the platforms but it's currently blocked by another Linux OTA issue